src/connectionAttempt.ts

```diff
@@ -0,0 +1,102 @@
+import { isCancelError, raceWithAbort } from '@mongodb-js/compass-utils';
+import type { ConnectionOptions, DataService } from 'mongodb-data-service';
+import { connect } from 'mongodb-data-service';
+
+import { createLogger } from './logging';
+import LINKS from './utils/links';
+
+// eslint-disable-next-line @typescript-eslint/no-var-requires
+const packageJSON = require('../package.json');
+
+const log = createLogger('connection attempt');
+
+function isConnectionAttemptTerminatedError(err: Error) {
+  return err?.name === 'MongoTopologyClosedError';
+}
+
+export class ConnectionAttempt {
```

Comment:
I was thinking about if it should be shared or not. Not having a place in Compass that really felt like it belonged was why I ended up with the copy paste. I feel keeping the `mongodb-data-service` package minimal is nice, but this could work there. With the slight logic change in error messaging its more apparent.
Could that be a follow up to this so that we can unblock the OIDC work for now?
Curious on other folks thoughts as well.

Reply:
@gribnoysup i recall you had some ideas here. `mongodb-data-service` feels appropriate for `ConnectionAttempt` as similarly to other places, this is where we make operations easily cancellable: https://github.com/mongodb-js/compass/blob/main/packages/data-service/src/data-service.ts#L1445C5-L1445C5

Regardless, I'm really not a fan of a copy-paste if this is one to be fair. Additionally, the need to tap into another package like `@mongodb-js/compass-utils` that i would expect hosting internals of Compass is another hint that we are possibly opening up the encapsulation and creating a coupling that we really can avoid.

Reply:
Going to open a pr in Compass soon and use it here when it's merged. If it's alright I'd like to merge this pr so I can work asynchronously on the OIDC pr, and merge in this refactor work, and avoid lots of merge conflicts. Do you feel that would be alright or is it better to wait?

Reply:
PR moving it to data-service so we can improve it here: https://github.com/mongodb-js/compass/pull/5261

Reply:
I don't remember really having any specific ideas around this, but generally speaking yeah, moving this to data service makes sense I think, I'll take a look at the PR 👍 And RE the promise cancellation and racing, IIRC the most (if not all) use cases we have for it are when working with data service methods, so yeah, eventually having this as part of the data service interface would be nice I think, but this is probably not something to do in the scope of the current taks

Reply:
Updated to use the connection attempt from mongodb-data-service.